### ISSUE TITLE
How to deal with compiled libraries

### ISSUE THREAD

**Issue by marcoromelli** [ubaniabalogun/serverless-package-python-functions#3]
I tried to use the plugin to package a python function with the `numpy` requirement. I get a successful deploy but when I try to call the lambda function the following error occurs:

> Importing the multiarray numpy extension module failed.  Most
> likely you are trying to import a failed build of numpy.
> If you're working with a numpy git repo, try `git clean -xdf` (removes all
> files not under version control).  Otherwise reinstall numpy.

Is there some fix for this?

**Comment by ubaniabalogun** [comment]
Hey @marcoromelli,
Thanks for the issue report. 
I've seen a solution for this problem that involves using Docker. I'll work on implementing it in the plugin but it will take some time to complete as I haven't spent much time using Docker. Pull requests are welcome.

In the meantime, a workaround that will allow you keep moving forward with your project is to get a version of numpy compiled for AWS Lambda and package it with the function using the function's `include` property (as opposed to specifying numpy in the requirements file). Here's an example below (I've used <----- to call out the important stuff):

Project structure

```
your-awesome-project/
├── compiled_libs
│   ├── numpy        <---- numpy compiled for AWS Lambda
├── function1
│   ├── lambda.py
│   └── requirements.txt      <--- your other Pip requirements
└── serverless.yml
```
And serverless.yml

```
service: your-awesome-project

plugins:
  - serverless-package-python-functions

custom:
  pkgPyFuncs: # plugin configuration
    buildDir: _build
    requirementsFile: 'requirements.txt'

functions:
  function1:
    name: function1
    handler: lambda.handler
    package:
      include:
        - ./function1
        - ./compiled_libs  <--- This will add numpy at the top level so "import numpy" still works
      artifact: ${self:custom.pkgPyFuncs.buildDir}/function1.zip
```

To get your hands on a version of numpy compiled for AWS Lambda, I believe you can spin up an EC2 host with the AMI described [here](http://docs.aws.amazon.com/lambda/latest/dg/current-supported-versions.html), `pip install numpy -t .` then copy the resulting files to your project.

[Docker Lambda](https://github.com/lambci/docker-lambda) may also provide an easier way to get the compiled library you need but I haven't worked with it yet. 

I hope this workaround lets you keep making progress on your project while a solution for this is worked on.
I'll update this issue ticket once the feature is in place.

**Comment by demone90** [comment]
Hello @ubaniabalogun 

nice workaround for this issue. Is there any chance to see soon the issue addressed in this project?
It would be really nice to have everything working and being this project one of the few working with python packaging and serverless it would even increase the popularity of it.

What do you mean with docker? Are you thinking about running a docker container with a very similar linux distro to the amazon servers env and compile the libs inside it? I could think about a possible merge request if I will find the time for doing it.

thank you so far for the nice plugin you built.

**Comment by ubaniabalogun** [comment]
Hey @demone90,
I started work on this today with plans for it to be available next week.

**Comment by ubaniabalogun** [comment]
@demone90 & @marcoromelli. 
Released version 0.2.0 with support for compiled libraries using Docker 🎉 
Instructions are on the docs page but the basics are
- Download and install Docker
- Add `useDocker: true` to the plugin configurations.
Give it a go and let me know what you think.
Closing this issue for now.
